packet/devices.go

```diff
@@ -0,0 +1,191 @@
+package packet
+
+import (
+	"context"
+	"strings"
+
+	"github.com/packethost/packngo"
+	"github.com/packethost/packngo/metadata"
+	"github.com/pkg/errors"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+)
+
+type instances struct {
+	client  *packngo.Client
+	project string
+}
+
+func newInstances(client *packngo.Client, projectID string) cloudprovider.Instances {
+	return &instances{client, projectID}
+}
+
+// NodeAddresses returns the addresses of the specified instance.
+func (i *instances) NodeAddresses(_ context.Context, name types.NodeName) ([]v1.NodeAddress, error) {
+	device, err := deviceByName(i.client, i.project, name)
+	if err != nil {
+		return nil, err
+	}
+
+	return nodeAddresses(device)
+}
+
+// NodeAddressesByProviderID returns the addresses of the specified instance.
+// The instance is specified using the providerID of the node. The
+// ProviderID is a unique identifier of the node. This will not be called
+// from the node whose nodeaddresses are being queried. i.e. local metadata
+// services cannot be used in this method to obtain nodeaddresses
+func (i *instances) NodeAddressesByProviderID(_ context.Context, providerID string) ([]v1.NodeAddress, error) {
+	device, err := i.deviceFromProviderID(providerID)
+	if err != nil {
+		return nil, err
+	}
+
+	return nodeAddresses(device)
+}
+
+func nodeAddresses(device *packngo.Device) ([]v1.NodeAddress, error) {
+	var addresses []v1.NodeAddress
+	addresses = append(addresses, v1.NodeAddress{Type: v1.NodeHostName, Address: device.Hostname})
+
+	var privateIP, publicIP string
+	for _, address := range device.Network {
+		if address.AddressFamily == int(metadata.IPv4) {
+			if address.Public {
+				publicIP = address.Address
+			} else {
+				privateIP = address.Address
+			}
+		}
+	}
+
+	if privateIP == "" {
+		return nil, errors.New("could not get private ip")
+	}
+	addresses = append(addresses, v1.NodeAddress{Type: v1.NodeInternalIP, Address: privateIP})
+
+	if publicIP == "" {
+		return nil, errors.New("could not get public ip")
+	}
+	addresses = append(addresses, v1.NodeAddress{Type: v1.NodeExternalIP, Address: publicIP})
+
+	return addresses, nil
+}
+
+// InstanceID returns the cloud provider ID of the node with the specified NodeName.
+// Note that if the instance does not exist or is no longer running, we must return ("", cloudprovider.InstanceNotFound)
+func (i *instances) InstanceID(_ context.Context, nodeName types.NodeName) (string, error) {
+	device, err := deviceByName(i.client, i.project, nodeName)
+	if err != nil {
+		return "", err
+	}
+
+	return device.ID, nil
```

Comment:
The providerID has to match the expected format, perhaps by changing this here to
```
   return fmt.Sprintf("packet://%s", device.ID), nil
```